[쇼핑 주문 3, 4단계] 페토 미션 제출합니다.

chanho0908 opened the pull request:
# 셀프 체크리스트

 <!--
 모든 항목을 점검한 후 리뷰를 요청하면, 보다 의미 있는 피드백을 받을 수 있습니다.
 -->

- [x] 프로그램이 정상적으로 작동하는가?
- [x] 모든 테스트가 통과하는가?
- [x] 이전에 받은 피드백을 모두 반영하였는가?
- [x] 코딩 스타일 가이드를 준수하였는가?
    - [x] IDE 코드 자동 정렬을 적용하였는가?
    - [x] 린트 검사를 통과하였는가?
- [x] 프로그래밍 요구 사항을 준수하였는가?
- [x] `README.md`에 기능 목록을 정리하고 명확히 기술하였는가?

# 어떤 부분에 집중하여 리뷰해야 할까요?

 <!--
 학습 목표에 맞게 구현하였는지 점검하고, 피드백을 받을 준비가 되었는지 확인하세요.
 정답을 묻기보다, 자신의 생각과 의사 결정 과정을 공유하며 피드백을 통해 논의하는 것이 좋습니다.
 -->
 
-

## 코드 리뷰 커뮤니케이션
### 1,2 단계 리뷰 
https://github.com/woowacourse/android-shopping-order/pull/100#discussion_r2123821562
>
```
buildConfigField(
            "String",
            "BASE_URL",
            "\"${properties.getProperty("base_url")}\"",
        )
    }
```
>질문을 바꿔볼게요.
코틀린 파일의 상수로 선언하지 않고 여기에 선언한 이유는 무엇인가요?
둘은 어떤 차이가 있을까요?
이곳에 배치하면 어떤 특징을 가져갈 수 있을까요?

코틀린 파일에 고정된 상수로 선언하는 것과 달리, `buildConfigField`를 사용하면 빌드 타입에 따라 다른 값을 넣을 수 있습니다.

예를 들어, 아래처럼 설정하면 `buildVariants`에서 debug 빌드를 선택하면 테스트 서버 URL이, release 빌드를 선택하면 실제 서버 URL이 자동으로 적용되어 각각 다른 환경에서 앱을 빌드할 수 있습니다.

```
buildTypes {
    debug {
        buildConfigField("String", "BASE_URL", "\"https://test.api.com\"")
    }
    release {
        buildConfigField("String", "BASE_URL", "\"https://prod.api.com\"")
    }
}

```

https://github.com/woowacourse/android-shopping-order/pull/100#discussion_r2123828782
https://github.com/woowacourse/android-shopping-order/pull/100#discussion_r2123835319

> 지금 이 파일만 봐도 enqueue에 대한 함수 커스텀이 3개나 들어있네요.
프로젝트가 점점 커진다면, 과연 3개로 끝날까요?
다른 개발자는 무엇을 선택해서 사용야할지 한 번에 알아낼 수 있을까요?

> 이 파일만 보고, 어떤 함수가 어떤 값을 반환하는지 알 수 있을까요?
내용물을 아는 사람에게는 아무런 값을 반환하지 않는 함수로 보이지만, 다른 개발자가 보았을 때도 동일하게 그렇게 보일지 고민해보시면 좋겠네요

말씀해주신 내용이 제가 생각하기에 비단 `enqueue`뿐만 아니라 개발을 하면서 항상 고민되는 내용인 것 같습니다 🤔

+ 같은 코드가 반복된다면 정말 중복인 것일까 ?
+ 중복이면 반드시 함수를 분리해야 하는가 ?
+ 중복이라고 함수를 분리했을 때 정말로 가독성이 좋아질까 ?
+ 함수가 짧아진다고 해서 좋은 코드인가 ?
+ 그럼 좋은 코드가 뭐지 ?

저도 이 코드를 분리하는 과정에서 말리빈이 말씀하신 것과 같은 고민을 했어요 !
> 다른 개발자는 무엇을 선택해서 사용야할지 한 번에 알아낼 수 있을까요?
말씀해주신 내용이 제 고민의 정답인 것 같네요 🙂

결국 함수를 분리하는 이유도, 중복을 제거하려는 이유도 코드를 개선하기 위해 했던 어떤 고민이든 좋은 코드를 만들기 위함이고,
좋은 코드란 같이 일하는 동료가 보기에 이해하기 쉬운 코드인 것 같아요 ! ☺️

코루틴을 적용하면서 해당 코드는 제거하게 되었지만 앞으로 개발하며 지금과 같은 고민을 할 때 말리빈이 말씀해주신
것들을 다시한번 리마인드 해보겠습니다 !

https://github.com/woowacourse/android-shopping-order/pull/100#discussion_r2123846996
> 로직으로만 보면 응답이 response.isSuccessful로 성공인데, Result.failure를 반환하는 이유는 잘 모르겠네요.
NPE인 이유는 무엇인가요?
실제로 이 Exception이 발생했을 때, 에러 메시지에 아무것도 나타나지 않는다면 디버깅하기 쉬운 코드일까요?

이 부분을 다시 고민해보니 무작정 `NullPointerException`로 처리한것이 얼마나 위험한지 깨닫게 되었습니다.
처음엔 응답이 `response.isSuccessful`로 성공해도 바디가 비어있다면, 당연 잘못된 값이 들어온 것으로 생각하였습니다.

하지만 다시 고민해보니 정말 순수하게 서버에 데이터가 비어있어 바디가 비어있을 수 있을 것 같아요.
예를 들어 처음 출시한 앱이면 댓글 같은데 당연히 없을 수 있으니까요 ! 🙃

https://github.com/woowacourse/android-shopping-order/pull/100#discussion_r2123861469
> Cart인 Product, Product인 Cart라면 어느 파일에 위치시켜야할까요?
도메인들이 점점 늘어나게된다면, 그 경우의 수들은 모두 어느 파일에 위치시켜야할까요?

으악 말씀해주신 내용이 정말 고민을 많이 했던 내용입니다 😵‍💫
사실 최근 본 상품도, 장바구니 상품도 결국 "상품"이라는 하나의 같은 도메인이 때문에 하나의 도메인으로 합쳐져도
충분히 합리적이라고 생각했습니다. 

다만, 이를 서로 다른 파일로 분리한 이유는 장바구니와 최근 본 상품에 대한 기능이 추가 되었을 때
이를 좀 더 독립적으로 관리하고자 함이었습니다. 

https://github.com/woowacourse/android-shopping-order/pull/100#discussion_r2123866073
> DataSource가 인터페이스가 아니었군요.
Repository와 DataSource는 어떨 때 사용하는걸까요?
어느 원천에서 온 데이터인지 이 이름만 보고는 알기가 너무 어렵네요.

엇..! 혹시 어떤 부분땜에 알기 어렵다고 하신건지 조금 더 알려주실 수 있으실까요 ?! 🥹
제가 생각하기엔 `CartDataSource `라는 class Type이 명시되어있어 알 수 있다고 생각해서요 ! ☺️

다만, 저도 `repository`와 `dataSource`를 사용하는 것에 대해 의문이 들었던 점이 하나 있는데요,
지금 `Repository`의 역할이 단순히 `DataSource`를 호출하는 것 뿐이라 `DataSource`없이 `repository`에서 직접 `retrofit service interface`를 주입받아 사용해도 된다고 생각해요 

그래서 이 부분에 대해 코치분들께 질문을 드리고 개인적으로 결론을 내봤습니다.

Repository의 역할은 데이터를 가지고 오는 원천을 추상화해 ViewModel과 데이터 소스(데이터를 저장하는 곳이라는 의미의 순수한 의미의 데이터 소스입니다 !)의 의존성을 분리하는 것이라고 생각합니다.

예를 들어 진구가 도라에몽한태 데이터 소스가 뭐야 ? 라고 물어봤을 때 도라에몽이 코치님들에게 질문을 해서 알아오던, 구글링을 해서 알아오던 진구는 도라에몽에게 필요한 정보를 전달받을 뿐, 그 정보를 어디서 가져왔는지 모르며 이 때 도라에몽이 Repository의 역할을 합니다

제가 생각하기에 `ViewModel`은(AAC)  `View`가 사용하는 `Model` 즉, 데이터를 복원하고 홀딩하는 홀더의 역할로써의 책임을 가지기 때문에 Ui 영역에 좀 더 밀접한 요소라고 생각해요. 때문에 데이터를 가져오는 곳과는 분리를 해야한다고 생각합니다.

그렇다면 Repository의 역할은 지금 충분히 존재 의의가 있는 것 같아요~! ☺️

그럼 DataSource는 무슨 일을 할까요 ? 

저는 Repository의 역할을 분리하는 아이라고 생각합니다. 

앞서 말씀드린 것 처럼 Repository는 이미 데이터 소스의 추상화라는 역할을 하고있어요. 그렇다면, 만약 특정 조건에 의해 Retrofit Service Interface에서 호출해야하는 API를 분기해야 한다면 ?

이미 Repository는 하나의 역할을 하고있을 뿐더러 때문에 데이터를 가져올 곳 까지 결정해야 한다면 `Repository Pattern`에서 말하는 `Repository`의 역할이 아니라고 생각해요. 

때문에 `Repository`는 데이터 소스를 추상화하고, 이 추상화된 데이터 소스에서 직접적으로 API를 호출해 데이터를 끌고 오는 아이가 바로 `DataSource`의 역할이라고 생각합니다.

https://github.com/woowacourse/android-shopping-order/pull/100#discussion_r2123883906
> postValue를 setValue로 바꾸면 어떤일이 발생할까요?
무슨 차이가 있을까요?
historyItems에서 가져온 값으로 setValue를하면 문제가 생기는데요, productRepository에서 가져온 값은 setValue를 해도 문제가 없어요. 왜일까요?

고것은 바로 historyItems를 가져오는 곳은 thread를 쓰고 있기 때문이죠 😎
```
@MainThread
protected void setValue(T value) {
    assertMainThread("setValue");
    mVersion++;
    mData = value;
    dispatchingValue(null);
}
```

`setValue`는 내부적으로 메인 스레드에서만 사용하도록 설계되어 있습니다. `assertMainThread`을 사용해 메인 스레드가 아닐 경우 다음과 같은 에러를 발생시킵니다

```
throw new IllegalStateException("Cannot invoke " + methodName + " on a background" + " thread");
```

`setValue`와는 다르게 postValue는 백그라운드 스레드에서 값을 쓸 수 있습니다. 다만, 값을 쓰는 행위 자체는 메인 스레드에서 하더군요 !

## 참고 자료

 <!-- 관련 문서나 참고할 만한 자료를 추가하세요. -->

## 스크린샷

 <!-- 필요하다면 스크린샷을 첨부하세요. -->


https://github.com/user-attachments/assets/32ec8f24-298d-4439-a96d-c46098bca257

chanho0908 committed:
feat : 최근 상품 목록 DB Dao Suspend 함수로 수정

chanho0908 committed:
test: 기존 테스트 제거

chanho0908 committed:
test: 테스트를 위한 FakeHistoryRepository, Dao, DataSource 추가

chanho0908 committed:
refactor: HistoryRepository 및 관련 클래스 코루틴 적용

- HistoryRepository의 함수들을 suspend 함수로 변경했습니다.
- 이에 따라 DefaultHistoryRepository의 스레드 처리 로직을 코루틴으로 변경했습니다.
- HistoryDataSource를 인터페이스로 변경하고, DefaultHistoryDataSource를 구현체로 추가했습니다.
- 테스트 코드 또한 코루틴 환경에 맞게 수정했습니다.

chanho0908 commented:
@coderabbitai pause

chanho0908 committed:
refactor: Local DB 장바구니 관련 파일 및 코드 삭제

chanho0908 committed:
refactor: `MockingServer` 및 관련 클래스 삭제

chanho0908 committed:
feat: 코루틴 테스트를 위한 CoroutinesTestExtension JUnit Jupiter Extension 추가

chanho0908 committed:
feat: 네트워크 오류 처리 기능 추가

- 네트워크 요청 결과 및 오류를 나타내는 `NetworkResult` 및 `NetworkError` 클래스를 정의했습니다.
- 네트워크 요청을 실행하고 결과를 처리하는 `NetworkResultHandler` 클래스를 추가했습니다.
- `NetworkResultHandler`는 `HttpException` 및 기타 예외를 로깅하고, 이를 `NetworkError`로 변환합니다.
- 네트워크 오류 발생 시 사용자에게 적절한 메시지를 토스트로 표시하는 `NetworkExceptionDelegator` 클래스를 추가했습니다.
- HTTP 오류 코드에 따른 사용자 메시지를 `strings.xml`에 추가했습니다.

chanho0908 committed:
feat: 상품 상세 화면 상품 정보 조회 로직 코루틴 적용 및 네트워크 에러처리 적용

chanho0908 committed:
feat: 상품 상세 화면 상품 정보 조회 로직 코루틴 적용 및 네트워크 에러처리 적용

chanho0908 committed:
refactor: ProductRepository, ProductService 비동기 처리 방식 변경

chanho0908 committed:
refactor: 상품 목록 화먄 네트워크 예외 처리 로직 위임

chanho0908 committed:
feat: `Context`에 `showToast` 확장 함수 추가

`messageResourceId`를 받아 토스트 메시지를 보여주는 `showToast` 확장 함수를 추가했습니다.

chanho0908 committed:
refactor: `CartDataSource.addCart`를 suspend 함수로 변경하고 `NetworkResult`를 반환하도록 수정

- `ApiCallbackHandler` 대신 `NetworkResultHandler`를 사용하도록 변경했습니다.
- Location 헤더가 비어있을 경우 `NetworkError.MissingLocationHeaderError`를 반환하는 테스트 코드를 추가했습니다.

chanho0908 committed:
refactor: `addCart` API 호출을 suspend 함수로 변경 및 Location 헤더에러 처리 추가

chanho0908 committed:
refactor: CartRepository의 loadSinglePage 콜백을 코루틴으로 변경

- DefaultCartRepository와 DataSource의 `loadSinglePage` 함수를 suspend 함수로 변경하고, `NetworkResult`를 직접 반환하도록 수정했습니다.
- ViewModel에서 `loadSinglePage` 호출 시 viewModelScope.launch를 사용하도록 변경했습니다.
- CartService의 `getCartSinglePage`를 suspend 함수로 변경하고, Call 대신 응답 객체를 직접 반환하도록 수정했습니다.

chanho0908 committed:
refactor: `CartRepository.updateQuantity`를 suspend 함수로 변경 및 콜백 제거

chanho0908 committed:
Refactor: 장바구니 상품 제거 코루틴 적용

chanho0908 committed:
Refactor: Remove CartEntity and update database version

chanho0908 committed:
Refactor: 최근 본 상품이 현재 상품으로 잘못 표시되는 문제 수정 및 이전 코드 제거

chanho0908 committed:
Refactor: 불필요한 CartSinglePage domain model 제거

chanho0908 committed:
Refactor: ApiCallbackHandler , BaseResponse 제거

chanho0908 committed:
Refactor: 상품 Repository IO Dispathcer 추가

chanho0908 committed:
Refactor: viewModel 코드 개선

chanho0908 committed:
Test:  MainViewModelTest 추가

chanho0908 committed:
feat: Retrofit 쿠폰 API 추가

chanho0908 committed:
feat: 쿠폰 도메인 모델 추가

- 쿠폰 종류를 나타내는 `Coupon` 인터페이스 정의
- 고정 할인 쿠폰 (`FixedCoupon`) 모델 추가
- 1+1 쿠폰 (`BogoCoupon`) 모델 추가
- 무료 배송 쿠폰 (`FreeshippingCoupon`) 모델 추가
- 타임세일 쿠폰 (`MiracleSaleCoupon`) 및 사용 가능 시간(`AvailableTime`) 모델 추가

chanho0908 committed:
feat: 쿠폰 도메인 모델 추가

- 쿠폰 종류를 나타내는 `Coupon` 인터페이스 정의
- 고정 할인 쿠폰 (`FixedCoupon`) 모델 추가
- 1+1 쿠폰 (`BogoCoupon`) 모델 추가
- 무료 배송 쿠폰 (`FreeshippingCoupon`) 모델 추가
- 타임세일 쿠폰 (`MiracleSaleCoupon`) 및 사용 가능 시간(`AvailableTime`) 모델 추가

chanho0908 committed:
feat: 쿠폰 도메인 모델 추가

- 쿠폰 종류를 나타내는 `Coupon` 인터페이스 정의
- 고정 할인 쿠폰 (`FixedCoupon`) 모델 추가
- 1+1 쿠폰 (`BogoCoupon`) 모델 추가
- 무료 배송 쿠폰 (`FreeshippingCoupon`) 모델 추가
- 타임세일 쿠폰 (`MiracleSaleCoupon`) 및 사용 가능 시간(`AvailableTime`) 모델 추가

chanho0908 committed:
feat: 주문 화면 RecyclerView Adapter 및 ViewHolder 구현

chanho0908 committed:
feat: 주문 화면 및 쿠폰 적용 기능 추가

chanho0908 committed:
feat: 가격 및 만료일 표시 형식 변경

- 가격 표시에 천 단위 구분 기호와 '원' 단위를 추가했습니다.
- 만료일 표시 형식을 'yyyy년 M월 d일'로 변경했습니다.
- 이를 위한 BindingAdapter를 추가했습니다.

chanho0908 committed:
feat: 주문 화면에 결제 버튼 추가

chanho0908 committed:
feat(MiracleSaleCoupon): 사용 가능 여부 확인 로직 추가

- `Coupon` 인터페이스에 만료 여부를 확인하는 `expired` 함수 추가
- `MiracleSaleCoupon`에 현재 시간 기준으로 쿠폰 사용 가능 여부를 판단하는 `isUsable` 함수 추가
- `isUsable` 함수는 만료일과 사용 가능 시간을 모두 고려하여 판단
- 관련 테스트 케이스 추가

chanho0908 committed:
feat: 최소 주문 금액 할인 쿠폰 사용 가능 여부 판단 로직 추가

chanho0908 committed:
test: 무료 배송 쿠폰 사용 가능 여부 테스트 케이스 추가

chanho0908 committed:
feat: BOGO 쿠폰 적용 가능 여부 및 적용 로직 추가

chanho0908 committed:
feat: Intent.getSerializableArrayList 확장 함수 추가

chanho0908 committed:
feat: 주문할 상품 정보를 주문 화면으로 전달

- 장바구니 목록과 추천 상품 목록에서 선택된 상품 정보를 주문 화면으로 전달
- 도메인 객체가 Serializable을 구현하도록 수정

chanho0908 committed:
fix(coupon): BOGO 쿠폰 적용 조건 수정

BOGO 쿠폰 사용 가능 조건에서 구매 수량이 '최소 구매 수량' 이상인 경우에서 '최소 구매 수량 + 증정 수량' 이상인 경우로 변경되었습니다.

chanho0908 committed:
refactor(CartViewModel): CartUiState.purchaseCart 사용하도록 변경

CartViewModel에서 화면 전환 시 전달하는 장바구니 목록을 CartUiState.purchaseCart를 사용하도록 수정했습니다.
이를 위해 CartUiState에 purchaseCart 프로퍼티를 추가하여 선택된 상품들의 Cart 목록을 반환하도록 변경했습니다.

chanho0908 committed:
feat: 주문 상품에 따라 사용 가능한 쿠폰만 필터링하는 기능 추가

- `ShoppingCart`에 결제 금액을 계산하는 `payment` 프로퍼티 추가
- `CouponValidator`를 구현하여 주문 상품 목록과 총 결제 금액을 기준으로 사용 가능한 쿠폰을 필터링하는 로직 추가
- `OrderViewModel`에서 `CouponValidator`를 사용하여 사용 가능한 쿠폰만 로드하도록 수정
- `OrderViewModelFactory`에 `CouponValidator` 주입 로직 추가

chanho0908 committed:
feat: 구매할 상품이 없는 경우 토스트 메시지 출력

장바구니 화면에서 구매할 상품을 선택하지 않고 '주문하기' 버튼을 누르면, 사용자에게 구매할 상품이 없다는 안내 메시지를 토스트로 보여줍니다.

chanho0908 committed:
fix: 미라클 세일 쿠폰 할인률 누락 수정

chanho0908 committed:
feat: Bogo 쿠폰의 기준 수량 계산 로직 추가

chanho0908 committed:
feat: 쿠폰 종류에 따른 할인 적용 로직 추가

chanho0908 committed:
feat: 주문 화면에서 쿠폰 적용 로직 추가

- 쿠폰 적용 시 주문 가격이 변경되도록 구현
- 쿠폰은 하나만 선택 가능하도록 구현

chanho0908 committed:
chore : 기타 코드 수정

chanho0908 committed:
feat: 주문 화면에서 네트워크 에러 처리 및 메인 화면으로 이동 기능 추가

chanho0908 committed:
feat: 주문 생성 기능 구현

- 주문 생성 API 연동
- 주문 완료 시 토스트 메시지 출력
- 주문 완료 시 메인 화면으로 이동

chanho0908 committed:
Merge branch 'step3' of https://github.com/chanho0908/android-shopping-order into step3

# Conflicts:
#	app/src/test/java/woowacourse/shopping/data/repository/DefaultCouponRepositoryTest.kt

chanho0908 committed:
test: 주문 화면 뷰모델 테스트 케이스 추가

chanho0908 committed:
feat: 의존성 주입을 위한 클래스 분리

chanho0908 committed:
refactor: TextViewAdapter 패키지 경로 변경

chanho0908 committed:
refactor: 코드 스타일 개선 및 `object`를 `data object`로 변경

chanho0908 commented:
@coderabbitai resume
@coderabbitai review

chanho0908 committed:
docs: 쿠폰, 상품 주문 기능 요구사항 명세 추가

chanho0908 committed:
refactor: CouponValidator에서 오늘 날짜를 LocalDateTime으로 생성 후 LocalDate로 변환

chanho0908 committed:
refactor: postValue to setValue for _uiEvent in MainViewModel

chanho0908 commented on app/src/main/java/woowacourse/shopping/domain/coupon/CouponApplierFactory.kt:
```diff
@@ -0,0 +1,48 @@
+package woowacourse.shopping.domain.coupon
+
```
이번 미션은 1단계 처럼 도메인에 관한 고민을 실컷 할 수 있어서 정말 재미있었던 것 같아요

물론 다른 요소들도 충분히 도메인적으로 풀 수 있겠지만 이미 너무 얽혀있어서 실타래를 다 풀어내기엔 미션 제출이 너무 늦어질 것 같아서 취사선택을 했습니다

대신 쿠폰에 대해선 정말 만을 고민하면서 최대한 비즈니스 로직을 도메인이 가질 수 있게 만들어봤습니다 !

malibinYun replied:
음... 솔직히 말씀드리면 이 객체가 무엇을 하는 역할인지 잘 이해가 안돼요.
어떤 일을 맡기고 싶었던 걸까요?

또, 쿠폰의 수는 기하급수적으로 늘어날텐데요, 거기에 유연하게 대처할 수 있는 코드인지도 가늠이 안되네요.
각 쿠폰마다 무언가 적용을 하는 메서드가 따로 존재해야하는 이유가 무엇인가요?

chanho0908 replied:
https://github.com/woowacourse/android-shopping-order/pull/122/commits/ed039fbb9a3e14d4488a563bb438a9ff0d281bcf

https://github.com/woowacourse/android-shopping-order/pull/122/commits/8c21533059a0d0f6477028abee0d6d016e9f4fc2

해당 리뷰 관련 커밋입니다 

다시 고민해보니 쿠폰의 유효성을 검사하고, 쿠폰을 적용하는 역할 모두 쿠폰 스스로가해야할 책임인 것 같아요 🙂 

지금 처럼 쿠폰의 유효성 검사 로직과 적용 로직이 분리되어있는 것은 응집도를 저하시키는 원인인 것 같습니다 

오랜만에 도메인적으로 문제를 풀어보려 하니 미흡한 요소가 많이 보이네요 레벨 1 때 했던 것들을 다시보며 리마인드 해야겠습니다 😭

chanho0908 commented on app/src/main/java/woowacourse/shopping/domain/exception/NetworkError.kt:
```diff
@@ -0,0 +1,19 @@
+package woowacourse.shopping.domain.exception
```
스웨거에 명시되어 있는 HttpException들은 예측 가능한 에러들로 생각해서
이를 커스텀해서 대응해봤습니다 !

chanho0908 commented:
@coderabbitai review

chanho0908 committed:
refactor: 장바구니 화면 네트워크 에러처리 미적용 수정

chanho0908 committed:
feat: MainActivity에 NetworkExceptionDelegator 미초기화 수정

chanho0908 committed:
fix: 미라클 세일 쿠폰 시간 계산 오류 수정

자정 이후까지 사용 가능한 미라클 세일 쿠폰의 경우, 시간이 만료되지 않았음에도 불구하고 만료되었다고 판단하는 오류를 수정했습니다.

chanho0908 committed:
refactor: 미라클 세일 쿠폰의 할인 금액에 -부호가 붙지 않는 오류 수정

chanho0908 committed:
refactor: 불필요한 변수 할당 제거, 오타 수정

chanho0908 committed:
refactor: 전체 페이지 로드를 위한 상수 정의

chanho0908 committed:
refactor: HistoryState를 Product 도메인 모델로 변경

- 기존 HistoryState를 Product 도메인 모델로 대체했습니다.
- 이에 따라 관련 클래스 및 레이아웃 파일의 타입 참조를 수정했습니다.

malibinYun commented on app/src/androidTest/java/woowacourse/shopping/data/db/HistoryDaoTest.kt:
```diff
@@ -33,37 +34,37 @@ class HistoryDaoTest {
-            historyDao.insert(it)
+            // then
+            val actual = historyDao.getLatestHistories()
+            assertEquals(actual, listOf(entity))
```
historyEntityFixture1을 저장하고 그대로 historyEntityFixture1를 비교하는 건 테스트로써 의미가 있을지 고민해보시면 좋겠어요.
[좋은 글](https://jojoldu.tistory.com/615) 하나를 소개해드릴게요 :)

chanho0908 replied:
헉 레벨 1때 첫 리뷰어였던 둘리께서 말씀해주셨던 자료네요 😱
레벨 1때 배웠던 것들을 잊지 않도록 다시금 리마인드 해야겠습니다 🥲

malibinYun commented on app/src/main/java/woowacourse/shopping/data/datasource/CartDataSource.kt:
```diff
@@ -1,33 +1,43 @@
     private val service: CartService,
-    private val handler: ApiCallbackHandler,
+    private val handler: NetworkResultHandler = NetworkResultHandler(),
 ) {
```
> 엇..! 혹시 어떤 부분땜에 알기 어렵다고 하신건지 조금 더 알려주실 수 있으실까요 ?! 🥹
제가 생각하기엔 CartDataSource 라는 class Type이 명시되어있어 알 수 있다고 생각해서요 ! ☺️

DataSource의 구현체는 어떤 원천에서 오는지를 접두사를 붙여 종종 사용하지요.
그런데 CartDataSource의 이름만 가지고는 Cart라는 데이터가 어디에서 오는지에 대한 원천을 알 길이 없네요 :)

chanho0908 replied:
https://github.com/woowacourse/android-shopping-order/pull/122/commits/17eb0913818a3a8626a9af96fe03a761d370d316

그런 의미였군요 !! 해당 코멘트 적용했습니다! ☺️

malibinYun commented on app/src/main/java/woowacourse/shopping/data/datasource/CartDataSource.kt:
```diff
@@ -1,33 +1,43 @@
-        callback: (Result<String>) -> Unit,
-    ) = handler.enqueueWithExtractHeaderLocationResult(service.addCart(request), callback)
+    suspend fun addCart(request: CartItemRequest): NetworkResult<Long> =
+        handler.execute {
```
다른 네트워크 호출을 부르는 함수를 추가해야할 때 handler.execute() 함수를 반드시 추가해야하는지 다른 개발자들이 알 수 있을지 고민해보시면 좋겠어요.
공통의 로직을 묶어서 객체 분리해보려는 시도는 아주 좋았어요.
좀 더 명확하게 역할을 수행하는지, 적절한 분리였는지 고민해보시면 좋겠네요 :)

게다가 execute란 이름은 Service 객체와 함께 있다보니 Retrofit의 execute함수를 연상하게되네요.

chanho0908 replied:
https://github.com/woowacourse/android-shopping-order/pull/122/commits/1bcef538b9b74da1c810257377dbef67e22c04b3

https://github.com/woowacourse/android-shopping-order/pull/122/commits/a33d9da2d576831cf252ccd22789812dabd07fed

해당 리뷰 관련 커밋입니다 !

이 부분에 대해서 제출 전까지 고민했습니다. 
>다른 네트워크 호출을 부르는 함수를 추가해야할 때 handler.execute() 함수를 반드시 추가해야하는지 다른 개발자들이 알 수 있을지 고민해보시면 좋겠어요.

제가 생각하기에 이 부분은 팀에서 컨벤션을 만들어 가이드 한다면 해소할 수 있는 문제가 아닐까 싶어요! 단순히 코드만 봤을 때 도 다른 코드에 모두 같은 로직이 들어가있으면 한번쯤은 들여다보지 않을까 하는 생각도 들더군요 🤔

다만, 제가 말씀드린 가능성이 존재하는 것 처럼 말리빈께서 말씀하신 경우의 수도 신중히 고려해야할 것 같습니다.
결과적으로 마지막까지 고민하다 에러를 전체적으로 다루는 함수는 제거하고, DataSource들의 함수에 모두
`runCatching-fold` 를 사용해 붙혀주고 최소한의 에러처리 부분만 유지했습니다

기존에 `getOrElse -> fold()`로 변경한 이유는 `fold`가 `getOrElse`에 비해서 성공과 실패의 케이스를 더 명확히 구분지어 인지할 수 있다고 생각해서 수정했습니다 !

malibinYun commented on app/src/main/java/woowacourse/shopping/domain/exception/NetworkError.kt:
```diff
@@ -0,0 +1,19 @@
+package woowacourse.shopping.domain.exception
+
+sealed class NetworkError : Throwable() {
+    data object UnknownError : NetworkError()
+
+    data object MissingLocationHeaderError : NetworkError()
+
+    sealed class HttpError : NetworkError() {
+        data object BadRequestError : HttpError()
+
+        data object AuthenticationError : HttpError()
+
+        data object AuthorizationError : HttpError()
+
+        data object NotFoundError : HttpError()
+
+        data object ServerError : HttpError()
+    }
+}
```
백엔드가 바뀌거나 아예 다른 서비스를 사용하는경우, 이 에러들이 반드시 발생할까요?
sealed로 묶여있으니, 이 안에서는 반드시 발생한다는 의미로 보이는데, 다른 것으로 교체했을 때 발생하지 않을 에러가 있을 수도, 새로운 에러 타입이 생길 수도 있겠어요.
유연하게 대처할 수 있을까요?

chanho0908 replied:
이 부분은 리뷰 제출시간이 얼마 남지 않아서 수정하지 못했습니다

백엔드가 바뀐다는 예시가 조금은 충격적인 것 같습니다 😱
물론 지금껏 열심히 디자인 패턴과 아키텍처를 적용한 이유가 외부 환경의 변화에 있어 변경을 최소화하기 위함도 있다고 생각해요 !
다만 너무 극단적인 예시처럼 느껴져 쉽사리 대응책이 떠오르지 않는 것 같아요 🥹

여기서 커스텀 에러 클래스를 만든 이유는 스웨거에 명시된 최소한의 에러에 대해 대응하고자 함입니다 🙂
다만 말씀해주신 것 처럼 sealed로 묶여있다보니 이 내부에 강하게 종속된 형태로 된 것 같아요

하지만 반대로 계층 구조를 만들어 관리할 수 있다는 장점도 있다고 생각합니다.

제겐 쉽사리 개선점들이 떠오르지 않아서 조금만 의견을 부탁드려도 될까요 !? ☺️

malibinYun replied:
그렇다면, Retrofit에서 이미 사용하고있는 HTTP Code 관련 객체들을 사용해보는 건 어떨까요?
네트워크 코드는 규격에 맞게 사용한다면 표준대로일테니까요.

이런 경우도 생각해보면 좋겠어요.
로그인시 실패하면 400을 내려줄 것인데, 서버는 아이디는 존재하는데 패스워드가 틀린 경우와, 아이디가 애초에 없는 경우를 구분하고 싶다면,
이것은 어떻게 구분해야할까요? 분명 401 네트워크 에러인데, sealed로 제한해뒀으니 HTTP 규약을 따르지 않는다고 백엔드에 일침을 날려야할까요? 아니면 표준에서 사용하지 않는 에러코드를 찾아내서 사용해야할까요? 그것이 나중에 표준에 등재될지도 모르지만요.
하지만 이 기능을 반드시 개발해야한다면, 유연하게 대처할 수 있는 형태일까요?

정답은 없어요. 실무는 늘 이런 정답이 없는 문제에서 싸워야한다는 점을 명심하세요 :)

chanho0908 replied:
이 부분에 대해서 정말 많은 고민을 해봤습니다.

스웨거에서 ID를 잘못 입력한 후 API를 요청해보니 실제로 다음과 같은 에러를 반환하더군요 !

<img width="351" alt="image" src="https://github.com/user-attachments/assets/1109749a-df4e-49f2-8106-7f82639a31f5" />

이를 대처하기 위해선 Response의 message 프로퍼티를 사용하면 될 것 같습니다 ☺️

다만, 말리빈이 제시해주신 문제점들을 이해하는데는 성공했으나 아직까지도 이 코드를 어떻게 수정해야할지 감이 잘 오지 않는 것 같아 기존 상태를 유지했습니다 😭

그래서 이 문제는 레벨 3에 프로젝트를 진행하며 실제 벡엔드 분들과 소통하고 함께 앱을 개발하게될 동료들과
함께 고민고 또 싸우며 문제의 답을 찾아나가 보겠습니다☺️

malibinYun replied:
굉장히 많은 전략들을 마주하게 될거예요.
실제 예시로, 실패를 하든 말든 무조건 일단 200으로 주는 백엔드도 왕왕 있답니다.
로그인 실패도 200이고 거기에 그들만의 에러 코드를 만들어 담거나, 에러 메시지를 담는 형태로요.
옛날 다녔던 회사에서 실제로 이런 전략을 사용했었답니다 허허

malibinYun commented on app/src/main/java/woowacourse/shopping/domain/exception/NetworkResult.kt:
기존의 Result 객체를 사용하지 않고 새로 만든 이유는 무엇인가요?

chanho0908 replied:
사실 Success, Failure 외에도 Loading 같은 상태를 추가하려고 만들었는데
시간에 쫓기다보니 구현하지 못했네요 😣

이 클래스는 제거했습니다 ! 😊

malibinYun commented on app/src/main/java/woowacourse/shopping/view/NetworkExceptionDelegator.kt:
이름만 보고는 마치 네트워크 예외 실행에 대한 대리자로 보이네요.
적절한 이름이었을까요?

chanho0908 replied:
https://github.com/woowacourse/android-shopping-order/pull/122/commits/4fa6902a9342fc6cf06ee3db7610475472684828

다소 오해를 부를만한 네이밍이네요  🥲
클래스명을 NetworkExceptionHandler로 변경해봤습니다 ㅎㅎ,,,

malibinYun commented on app/src/main/java/woowacourse/shopping/data/di/RepositoryModule.kt:
```diff
@@ -0,0 +1,35 @@
+
+    val defaultProductRepository: ProductRepository by lazy {
+        DefaultProductRepository(dataSourceModule.productsDataSource)
+    }
```
DefaultCouponRepository 만 구현체 타입을 선언한 이유는 무엇인가요?

chanho0908 replied:
으악 실수입니다!!  😣

malibinYun commented on app/src/main/java/woowacourse/shopping/data/network/response/coupons/CouponsResponse.kt:
쿠폰 응답 객체 자체를 sealed로 받아보는 건 어떨까요?
역직렬화/직렬화 라이브러리를 사용하니, 어떤 기능을 사용할 수 있을지 직접 찾아보세요 :)

chanho0908 replied:
https://github.com/woowacourse/android-shopping-order/pull/122/commits/626f20702a5524e9282948c99e1ce0fd4ca6a188

해당 리뷰 관련 커밋입니다 !

malibinYun commented on app/src/main/java/woowacourse/shopping/data/network/service/CartService.kt:
```diff
@@ -13,24 +13,24 @@ import woowacourse.shopping.data.network.response.carts.CartsResponse
 interface CartService {
     @POST("/cart-items")
-    fun addCart(
+    suspend fun addCart(
         @Body request: CartItemRequest,
-    ): Call<Unit>
+    ): Response<Unit>
```
어느 곳은 Response\<Unit> 를 반환하고, 어느 곳은 아무것도 반환하지 않은 이유는 무엇인가요?

chanho0908 replied:
Response<Unit>을 반환하는 곳은 Header에서 Location을 통해
장바구니 상품 아이디를 추출하기 위함입니다 ! ☺️

malibinYun replied:
이 파일만 보면, 다른 개발자들은 반환 타입을 어떻게 해야할지 혼동이 올거라 생각해요.
헤더를 사용한다는 사실을 모를테니까요.
코드의 사용을 통일해보는 건 어떨까요?

chanho0908 replied:
https://github.com/woowacourse/android-shopping-order/pull/122/commits/564ec434ce7a405f13514107ead2ae51e5c03938

해당 내용 관련 커밋입니다 !

malibinYun commented on app/src/main/java/woowacourse/shopping/data/repository/DefaultCartRepository.kt:
```diff
@@ -1,42 +1,41 @@
-        dataSource.singlePage(page, pageSize) { callback(it) }
-    }
+    ): NetworkResult<CartsSinglePage> =
+        withContext(Dispatchers.IO) {
+            cartDataSource.singlePage(page, pageSize)
+        }
```
withContext를 사용한 이유는 무엇인가요?

사용하지 않아도 아무런 문제가 발생하지 않는데요, 그 이유가 무엇인지에 대해 엮어 고민해보세요 :)

chanho0908 replied:
`suspend` 함수는 Kotlin 컴파일러에 의해 내부적으로 `Continuation` 객체를 인자로 받는 형태로 변환됩니다. 이 `Continuation`은 중단 지점(suspension point)을 기억하고, 이후 `resumeWith(...)`을 통해 실행이 재개됩니다.

`Retrofit`은 `suspend` 함수 호출 시 인자로 전달된 `Continuation`을 감지하고, 내부적으로 스레드를 전환합니다.

때문에 `ViewModelScope.launch` 가 메인 스레드(`Main.immediate`)에서 동작해도 문제가 발생하지 않습니다.

하지만 Room의 경우 별도로 IO 스레드로의 전환을 지원하지 않더군요 ! 🤔
```
// 1안
class FooRepository(){
    suspend fun doSomething() { // ... }
}

viewModelScope.launch(Dispatchers.IO) {
  val result = fooRepository.doSomething(...)
  withContext(Dispatchers.Main){
       liveData.value = result
   }
}

// 2안
class FooRepository(){
    suspend fun doSomething() = withContext(Dispatchers.IO){ // ... }
}
viewModelScope.launch{
  val result = fooRepository.doSomething(...)
  liveData.value = result
}
```

그래서 위와 같이 두 가지 경우의 수 중 선택을 해야했는데, 

스레드 전환은 호출부(ViewModel)가 아니라, 실제 I/O가 필요한 지점(Repository)에서 처리하도록 해야한다고 생각했어요 !

다시 돌아와서 Retrofit 이야기를 해보겠습니다

Retrofit이 내부적으로 백그라운드 스레드 전환을 해주는 것 또한 맞지만 

Repository가 withContext(Dispatchers.IO)를 직접 사용하면,
그 메서드는 semantic하게 "I/O가 필요한 작업"임을 보여준다고 생각해요 !

Retrofit이 백그라운드 전환을 자동으로 해준다고 해도

처음 보는 사람이 이 메서드가 I/O를 수행하는지, 스레드 전환이 필요한지 바로 알 수 없을 것 같습니다 🤔

그래서 저는 `Repository`에서 명시적으로 `withContext(Dispatchers.IO)`를 사용했습니다

malibinYun commented on app/src/main/java/woowacourse/shopping/domain/coupon/CouponValidator.kt:
```diff
@@ -0,0 +1,25 @@
+import java.time.LocalDateTime
+
+class CouponValidator : CouponValidate {
+    override fun validCoupon(
+        coupons: List<Coupon>,
+        orders: List<ShoppingCart>,
+    ): List<Coupon> {
+        val now = LocalDateTime.now()
+        val today = now.toLocalDate()
+        val totalPayment = orders.sumOf { it.payment }
```
테스트 가능한 형태일까요?

chanho0908 replied:
https://github.com/woowacourse/android-shopping-order/pull/122/commits/052a8ef6c21759fba643e1d50f433b67a81c0ccb

해당 리뷰 관련 커밋입니다 !

사실 `Validator` 내부에서 사용되는 검증 로직을 모두 테스트하였기 때문에 `Validator` 자체는 테스트하지 않아도 된다고 생각해서 
직접 생각하게 했었습니다 😅

chanho0908 committed:
feat: 장바구니에 담긴 상품 수량 조회 기능 추가

- CartRepository에 장바구니 상품 수량 조회 메서드 추가
- CartDataSource에 장바구니 상품 수량 조회 메서드 추가 및 API 호출 로직 구현
- CartCountResponse 데이터 클래스 추가
- CartService에 장바구니 상품 수량 조회 API 엔드포인트 추가

malibinYun commented on app/src/main/java/woowacourse/shopping/view/order/state/OrderUiState.kt:
```diff
@@ -0,0 +1,80 @@
+        return copy(coupons = updatedCoupons, payment = newPayment)
+    }
+
+    private fun applyCoupon(
+        coupon: Coupon,
+        couponFactory: CouponApplierFactory,
+    ): Payment {
+        return when (coupon) {
+            is BogoCoupon -> couponFactory.bogoCouponApplier.apply(originPayment, order, coupon)
+            is FixedCoupon -> couponFactory.fixedCouponApplier.apply(originPayment, coupon)
+            is FreeshippingCoupon -> couponFactory.freeShippingCouponApplier.apply(originPayment)
+            is MiracleSaleCoupon -> couponFactory.miracleSaleCouponApplier.apply(originPayment, coupon)
+        }
+    }
```
쿠폰 적용이 필요한 곳에서 매번 이런 코드를 써주어야할까요?
쿠폰의 개수가 기하급수적으로 늘어난다면, 유연하게 대처 가능한 형태일까요?

chanho0908 replied:
https://github.com/woowacourse/android-shopping-order/pull/122/commits/5dea2a9c4bca0db4135a5f2075a62865b1005d9e

해당 리뷰 관련 커밋입니다 !

쿠폰 종류에 따른 분기에 대한 책임을 CouponApplierFactory이 스스로하고 
OrderUiState에선 이를 호출하기만 하도록 수정했습니다 !

malibinYun reviewed (changes requested):
제가 시간을 착각해서 리뷰가 늦어졌네요 죄송해요. 😥
1, 2단계 미션 고생 많으셨어요.
여러 고민해볼 만한 곳에 코멘트 남겨두었으니 확인해주세요.
피드백 반영 후 다시 리뷰 요청 부탁드려요 :)

chanho0908 committed:
refactor: 장바구니 수량 변경 시 메인 화면에 반영 및 전체적인 로직 수정

- 장바구니 상품 추가, 수량 변경, 삭제 시 메인 화면의 장바구니 아이콘에 수량 즉시 반영
- 상품 UI 상태에서 cartId가 null인 경우를 `NOT_IN_CART` 상수로 명확히 표현
- 최근 본 상품 목록 데이터 타입을 Product로 변경
- MainViewModel에서 LiveData 조합 방식을 MediatorLiveData에서 직접 관리로 변경
- 기타 코드 포맷팅 및 리팩토링

malibinYun commented:
> 하지만 다시 고민해보니 정말 순수하게 서버에 데이터가 비어있어 바디가 비어있을 수 있을 것 같아요.
예를 들어 처음 출시한 앱이면 댓글 같은데 당연히 없을 수 있으니까요 ! 🙃

바디가 비어있는 건 예시로 들어주신 경우는 아닐거라 생각해요. 정말 데이터가 없으면 [] 같은 빈 제이슨 배열을 주는 게 보통이니까요 :)

> 으악 말씀해주신 내용이 정말 고민을 많이 했던 내용입니다 😵‍💫
사실 최근 본 상품도, 장바구니 상품도 결국 "상품"이라는 하나의 같은 도메인이 때문에 하나의 도메인으로 합쳐져도
충분히 합리적이라고 생각했습니다.
>
> 다만, 이를 서로 다른 파일로 분리한 이유는 장바구니와 최근 본 상품에 대한 기능이 추가 되었을 때
이를 좀 더 독립적으로 관리하고자 함이었습니다.

Service 파일을 통째로 하나로 관리하는 건 어떤가요?

> 앞서 말씀드린 것 처럼 Repository는 이미 데이터 소스의 추상화라는 역할을 하고있어요. 

네. 맞아요.
그러나 데이터 소스의 구현체에 그래서 어디서 오는 것인지에 대한 명칭이 없다면, 다른 개발자들이 내부를 보지 않고도 서버를 거치는 것인지, 로컬을 거치는 것인지 알 수 있을까요?

> setValue와는 다르게 postValue는 백그라운드 스레드에서 값을 쓸 수 있습니다. 다만, 값을 쓰는 행위 자체는 메인 스레드에서 하더군요 !

네 맞아요.
그럼 다른 개발자들이 setValue를 써야할 지, postValue를 써야할 지, 내 코드의 상세 구현을 보지 않고 한 번에 알아낼 수 있을까요?
레트로핏으로 가져온 데이터들은 왜 setValue를 사용하더라도 아무런 문제가 생기지 않을까요?

chanho0908 committed:
refactor: remote, local dataSource 분리

chanho0908 committed:
refactor: CouponRepository 추상화 적용

chanho0908 committed:
refactor: NetworkExceptionDelegator 클래스명을 NetworkExceptionHandler로 변경

chanho0908 committed:
refactor: 네트워크 요청 결과 처리 함수 수정

chanho0908 committed:
refactor: 네트워크 요청 결과 처리 함수 수정 적용

chanho0908 committed:
refactor: `isExpired` 로직을 `LocalDate.isBefore`를 사용하도록 변경

chanho0908 committed:
refactor: 쿠폰 response class를 sealed interfacle로 변경

chanho0908 committed:
feat: `ShoppingCarts` 일급 컬렉션 추가

chanho0908 committed:
refactor: `getSerializableArrayList`를 `getSerializableObject`로 변경

chanho0908 committed:
refactor: CouponApplierFactory에서 쿠폰 적용 로직 분리 및 BogoCoupon 로직 수정

- CouponApplierFactory에서 각 쿠폰 타입에 대한 apply 로직을 private 메서드로 분리했습니다.
- BogoCoupon 적용 시, buyQuantity와 getQuantity를 더한 값으로 수량을 확인하도록 수정했습니다.
- DetailViewModel에서 실패 처리 로직 주석을 해제했습니다.
- ShoppingCarts에서 쿠폰 적용 가능한 상품이 없을 경우 예외를 발생시키도록 변경했습니다.

chanho0908 committed:
refactor: CouponValidate 인터페이스에 현재 시간 파라미터 추가

chanho0908 committed:
fix: 장바구니 수량 변경 후 총 수량 갱신

chanho0908 committed:
refactor: 장바구니 페이징 사이즈 오류 해결 20개 -> 5개로 수정

chanho0908 committed:
refactor: 데이터 소스 이름 변경

- DefaultHistoryDataSource -> LocalHistoryDataSource
- ProductDataSource -> RemoteProductDataSource

chanho0908 committed:
refactor: `NetworkResultHandler`를 예외 핸들링에만 사용하도록 수정

- `NetworkResultHandler`의 `handleResult` 메서드를 제거하고, 각 DataSource에서 `runCatching`을 사용하도록 변경했습니다.
- `handleException` 메서드는 기존의 예외 처리 로직을 유지하며, 네트워크 오류 및 기타 예외를 적절한 `NetworkError`로 매핑합니다.
- `CouponValidate` 및 `CouponValidator` 인터페이스의 `validCoupon` 메서드 파라미터명 `now`의 불필요한 공백을 제거했습니다.
- `NetworkResultHandlerTest`의 테스트 케이스를 변경된 로직에 맞게 수정했습니다.

chanho0908 committed:
test: HistoryDaoTest 예상값 하드코딩

chanho0908 commented:
@coderabbitai review

malibinYun commented on app/src/main/java/woowacourse/shopping/data/network/service/CouponService.kt:
>> 으악 말씀해주신 내용이 정말 고민을 많이 했던 내용입니다 😵‍💫
사실 최근 본 상품도, 장바구니 상품도 결국 "상품"이라는 하나의 같은 도메인이 때문에 하나의 도메인으로 합쳐져도
충분히 합리적이라고 생각했습니다.
>>
>> 다만, 이를 서로 다른 파일로 분리한 이유는 장바구니와 최근 본 상품에 대한 기능이 추가 되었을 때
이를 좀 더 독립적으로 관리하고자 함이었습니다.
>>
> Service 파일을 통째로 하나로 관리하는 건 어떤가요?

위 부분에 대해서 고민을 많이 해보셨나요?

chanho0908 replied:
장바구니, 상품, 최근 본 상품은 서로 다른 비즈니스 로직과 API를 갖고 있기 때문에 이를 각각 별도의 서비스 파일로 관리하면 각 기능의 책임이 명확해진다고 생각해요 !

만약 기능이 추가되어 API가 많아지더라도 한 기능에만 집중해 수정할 수 있을 것 같습니다 🤔

또한 각 기능별로 담당하는 사람 또는 팀이 다를 경우, 기능별로 파일이 분리되어 있으면 자신이 담당하는 부분만 사용하면 되니 예상치 못한 수정이나 오용을 방지할 수 있는 이점도 있을 것 같습니다

malibinYun commented on app/src/main/java/woowacourse/shopping/data/repository/DefaultCartRepository.kt:
```diff
@@ -1,42 +1,45 @@
         page: Int?,
         pageSize: Int?,
-        callback: (Result<CartsSinglePage>) -> Unit,
-    ) {
-        dataSource.singlePage(page, pageSize) { callback(it) }
-    }
+    ): Result<CartsSinglePage> =
+        withContext(Dispatchers.IO) {
+            remoteCartDataSource.singlePage(page, pageSize)
+        }
```
https://github.com/woowacourse/android-shopping-order/pull/122#discussion_r2135418003

허허.. 

> 처음 보는 사람이 이 메서드가 I/O를 수행하는지, 스레드 전환이 필요한지 바로 알 수 없을 것 같습니다 🤔

이것은 다른 사람이 내 loadSinglePage() 메서드를 보아도 똑같이 드는 생각이겠어요. 명시적으로 사용하더라도요.
게다가 이미 처리를 해둔 코드를 또다시 context를 옮기는 것은 불필요한 작업이라 생각해요.

레트로핏은 왜 외부에서 스레드 변환이 있었는지 알지도 못하게 자기네들 마음대로 만들어둔걸까요?
이건 레트로핏 만의 것이 아니랍니다.

**"suspend" 키워드가 붙었다 해서, 이 함수가 백그라운드에서 실행한다는 의미가 아니다.**
https://medium.com/androiddevelopers/coroutines-on-android-part-i-getting-the-background-3e0e54d20bb
위 글을 참고해보세요. 그리고 레트로핏의 반환 값을 메인스레드에서 사용했을 때 문제가 없던 이유도 같이 엮어서 생각해보세요 :)

문서를 잘 찾아보시면 안드로이드 권장 사항 문서에도 닿을 수 있겠어요.

chanho0908 replied:
말리빈 .. 원피스를 찾은 것 같습니다

https://developer.android.com/kotlin/coroutines/coroutines-best-practices#main-safe

해당 문서를 참고하고 정리해봤습니다 !

>Suspend 함수는 반드시 메인 스레드의 호출로부터 안전해야 한다

네트워크 API 호출 같은 무거운 작업을 메인 스레드에서 처리하면 앱이 멈추거나 ANR이 발생하기 때문에
무거운 작업은 백그라운드에서 수행하고, 결과만 **메인 스레드로 다시 전달**하는 방식이 필요합니다.

레트로핏이 어떻게 내부적으로 스레드 전환을 하는지 찾아보았습니다.

[해당 자료](https://www.youtube.com/watch?v=x0KY6qtuNHg&t=1741s)를 보면 이번 미션에서 
코루틴을 적용하기 전에 사용했던 enqueue를 사용해 백그라운드에서 동작하도록 구현되어있군요 !

저는 코루틴의 IO Dispatcher로 전환된다고 생각했는데 아니었네요 😅

정리해보면, 
1. `ViewModelScope.lauch`를 사용해 메인 스레드에서 API 호출
2. enqueue를 사용해 OkHttp의 내부 스레드 풀에서 비동기적으로 콜백을 처리
3. 결과는 코루틴의 재개 지점(Continuation)에 resumeWith(Result<T>) 형태로 넘김
5.  코루틴 재개 
6.  원래의 코루틴 컨텍스트(예: Dispatchers.Main)에 따라 계속 실행
 
이 플로우로 흘러가기 때문에 공식 문서에서 말하는 `main thread`로 부터 안전한 `suspend` 함수가 되는 것 같습니다

말씀해주신 "suspend" 키워드가 붙었다 해서, 이 함수가 백그라운드에서 실행한다는 의미가 아니다.는
suspend 함수는 단지 일시 중단이 가능하다는 의미이지, 백그라운드 스레드에서 실행하는 것을 보장하지 않습니다.

하지만 레트로핏의 경우, CallBack 방식의 enquque를 사용해 백그라운드 스레드에서 처리되기 때문에
해당 [디스커션](https://github.com/square/retrofit/discussions/4024)에 의하면 따로 Dispatcher를 지정해주지 않아도 되는 것으로 해석되는 것 같습니다 ..!

malibinYun commented on app/src/main/java/woowacourse/shopping/domain/cart/ShoppingCarts.kt:
```diff
@@ -0,0 +1,30 @@
+        get() = shoppingCarts.map { it.id }
+
+    val totalPayment: Int
+        get() = shoppingCarts.sumOf { it.payment }
```
이 값들은 참조 될 때마다 연산을 하겠어요. 그럴 필요가 있을까요?
불필요한 연산을 줄이는 방법은 무엇이 있을까요?

chanho0908 replied:
https://github.com/woowacourse/android-shopping-order/pull/122/commits/2aa38bec668513a6436989f38d381536ccaf3cfb

해당 리뷰 관련 커밋입니다 !

값을 매번 계산할 필요가 없는 것들은 초기화 시점에 한번만 계산하도록 프로퍼티로 선언했습니다 !

malibinYun commented on app/src/main/java/woowacourse/shopping/domain/coupon/CouponValidator.kt:
```diff
@@ -0,0 +1,24 @@
+                is FreeshippingCoupon -> coupon.isUsable(today, totalPayment)
+                is MiracleSaleCoupon -> coupon.isUsable(now)
+            }
+        }
```
이 검증기가 필요한 이유가 와닿지 않네요.
하는 역할을 보면 검증기라기보단 필터기에 가깝다는 생각이 더 많이 드네요.
정말 필요한 역할의 객체였나요?

chanho0908 replied:
필터... 맞습니다,,, 😂 검증 로직은 쿠폰 스스로가 가지도록 수정했습니다

malibinYun commented on app/src/main/java/woowacourse/shopping/domain/coupon/FixedCoupon.kt:
```diff
@@ -0,0 +1,24 @@
+    ): Boolean {
+        if (isExpired(standardDate)) return false
+        if (!isMinimumAmountSatisfied(standardAmount)) return false
+        return true
```
standardAmount가 어떤 뜻인지 짐작이 잘 가지 않네요.
또, isUsable은 모든 쿠폰에서 가지고 있는데, interface의 공통 메서드로 분리해보는건 어떨까요?

또 다른 관점으로는, isUsable이란 의미는 굉장히 범위가 큰 역할은 아닐지 고민해보셔도 좋겠어요.

malibinYun commented on app/src/main/java/woowacourse/shopping/view/cart/vm/CartViewModel.kt:
```diff
@@ -46,45 +50,46 @@ class CartViewModel(
                                     state.modifyUiState(updated.copy(cartId = cartId))
                                 _cartUiState.value = _cartUiState.value?.addCart(cartId, updated)
                             }
                             .onFailure(::handleFailure)
```
객체지향 생활체조의 인덴트 규칙을 잊으신 건 아니겠죠? 늘 지켜보려고 노력해보세요 :)

malibinYun commented on app/src/main/java/woowacourse/shopping/view/core/bindingadapter/TextViewAdapter.kt:
```diff
@@ -0,0 +1,25 @@
+fun TextView.setPriceText(price: Int) {
+    val pattern = context.getString(R.string.thousand_unit_format)
+    text = pattern.format(price)
+}
```
파라미터 입력 값이 null로 들어오는 경우는 없을까요?

chanho0908 replied:
https://github.com/woowacourse/android-shopping-order/pull/122/commits/c203797b42e4b1a6cdc3d61b134d6e5158bc19bb

해당 내용 관련 커밋입니다 !

malibinYun reviewed (changes requested):
오늘 개인적인 일이 많아서 리뷰가 늦었네요 죄송합니다.
여러 생각해볼 곳들에 코멘트 남겨두었으니 깊게 고민해보는 시간 가져보세요.
다른 질문들에도 답변 남겨주시면 너무 좋겠어요 :)
피드백 반영 후 다시 리뷰 요청 부탁드려요!

chanho0908 committed:
refactor: CouponApplierFactory, CouponValidator, CouponValidate 제거

chanho0908 committed:
refactor: 쿠폰 적용 및 사용 가능 여부 확인 로직 변경

- Coupon 인터페이스에 isUsable, applyToPayment 메서드 추가
- 각 쿠폰 클래스에서 isUsable, applyToPayment 메서드 구현
- isUsable 메서드에서 주문 정보와 결제 금액을 인자로 받도록 변경
- BogoCoupon의 applyToPayment 로직 추가
- MiracleSaleCoupon의 applyToPayment 로직 추가
- FixedCoupon의 applyToPayment 로직 추가
- FreeshippingCoupon의 applyToPayment 로직 추가
- 관련 테스트 코드 수정

chanho0908 committed:
refactor: `NetworkExceptionHandler`를 core/handler 패키지로 이동

chanho0908 committed:
refactor: ViewModel에서 cart 관련 로직 분리

chanho0908 committed:
refactor: TextViewAdapter null 처리

chanho0908 committed:
refactor: 매번 계산이 필요하지 않은 프로퍼티들의 custom getter 제거

chanho0908 committed:
refactor : 쿠폰 ui 상태 수정 함수 개선

chanho0908 committed:
refactor: Retrofit 서비스 함수 반환 타입을 Response<Unit>으로 변경

기존 Unit 타입에서 Response<Unit>으로 변경하여 서버 응답 코드를 확인할 수 있도록 수정합니다.

chanho0908 reviewed (commented):
안녕하세요 말리빈!

오늘 처음 뵙게 되어 너무 반갑고 영광이었습니다 ☺️
누가 말리빈 보고 “성능 아저씨”라던데 ㅎㅎㅎ

물론! 말리빈의 말씀처럼 정답이란 없겠지만, 정말 아쉽게도 오늘 질문드린 내용에 대한 더 나은 방법은 아직 찾지 못했습니다 😭
아직까지 실제 서비스 환경에서 문제를 마주한 경험이 없다 보니, 제 스스로도 납득할 만큼의 결과물이 나오지 않아 현상을 유지했어요…

말리빈께서 ‘정답은 없다’라고 말씀주신 것을 듣고 리뷰를 다시 돌아보니,
스스로 명확한 의도와 목적을 가지고 구현하였는가? 라는 질문을 던질 수 있는 여러 요소들에 귀신같이 리뷰를 남겨주신 것 같아요.

이번 기회에 그 부분을 깊이 고민해보고, 앞으로도 더 명확한 의도와 목적을 갖고 코드를 작성하는 데 큰 도움이 될 것 같습니다.

감사합니다 😊

chanho0908 committed:
refactor: repository withContext 제거 implementations

malibinYun commented on app/src/main/java/woowacourse/shopping/data/datasource/local/HistoryDataSource.kt:
```diff
@@ -0,0 +1,9 @@
+
+import woowacourse.shopping.data.db.entity.HistoryEntity
+
+interface HistoryDataSource {
+    suspend fun latestHistory(): List<HistoryEntity>
+
+    suspend fun insertHistory(entity: HistoryEntity)
+}
```
History라는 이름만 있으면, 앞으로 추가될 비슷한 의미를 갖는 다른 도메인을 나타내기 곤란해진다 생각해요.
더 구체적인 단어를 사용해보시면 좋겠어요 :)

malibinYun commented on app/src/main/java/woowacourse/shopping/data/datasource/remote/RemoteCartDataSource.kt:
```diff
@@ -0,0 +1,65 @@
+    private val handler: NetworkResultHandler,
+) {
+    suspend fun addCart(request: CartItemRequest): Result<Long> =
+        runCatching {
+            val response = service.addCart(request)
+            val location =
+                response.headers()["Location"]
+                    ?.substringAfterLast("/")
+                    ?.toLongOrNull()
+            requireNotNull(location) { throw NetworkError.MissingLocationHeaderError }
+            location
```
runCatching 내의 구문이 길어지면,
그래서 어느 부분에서 exception이 나길래 씌워져있을까 하는 의문점을 남긴다 생각해요.
runCatching이 필요한 부분만 국소적으로 사용해보는 건 어떨까요?

malibinYun commented on app/src/main/java/woowacourse/shopping/data/di/RetrofitModule.kt:
```diff
@@ -9,11 +9,10 @@ import okhttp3.logging.HttpLoggingInterceptor
+import woowacourse.shopping.data.network.BasicAuthentication
 
-object RetrofitProvider {
-    private const val BASE_URL = BuildConfig.BASE_URL
+class RetrofitModule {
```
이친구는 class가 된 이유가 궁금하네요~
object였을 때와 어떤 차이점이 있었나요?

malibinYun commented on app/src/main/java/woowacourse/shopping/domain/cart/ShoppingCarts.kt:
```diff
@@ -0,0 +1,28 @@
+            .maxByOrNull { it.product.priceValue }
+            ?.product
+            ?.priceValue
+            ?: throw IllegalArgumentException(CAN_NOT_FOUND_PRODUCT)
+    }
```
어떤 값을 찾는 메서드이네요.
그런데 잘못된 값을 줬기 때문에 못찾은 것인데, exception이 발생하는 건 자연스러운 일일까요?
Map의 get메서드, find 확장함수 등등.. 다른 곳에서는 찾는 메서드에 못찾았을 때 어떻게 수행하고 있는지 들여다보셔도 좋겠어요.

malibinYun commented on app/src/main/java/woowacourse/shopping/domain/coupon/FreeshippingCoupon.kt:
```diff
@@ -0,0 +1,38 @@
+        )
+    }
+
+    private fun isMinimumAmountSatisfied(standardAmount: Int) = standardAmount >= minimumAmount
+}
```
함수 반환 타입은 항상 써주는 건 어떨까요?
컴파일러야 자동 추론해서 잘 알겠지만, 다른 개발자들은 이 함수가 어떤 값을 반환하는지 즉시 알 수 있을까요?

malibinYun commented on app/src/main/java/woowacourse/shopping/view/core/handler/NetworkExceptionHandler.kt:
```diff
@@ -0,0 +1,43 @@
+
+    private fun asResourceId(error: Throwable): Int {
+        return when (error) {
+            NetworkError.HttpError.AuthenticationError -> {
+                R.string.http_error_authentication_message
+            }
+            NetworkError.HttpError.AuthorizationError -> {
+                R.string.http_error_authorization_message
+            }
+            NetworkError.HttpError.BadRequestError -> {
+                R.string.http_error_bad_request_message
+            }
```
HttpError에 대한 확장 프로퍼티로 R값을 반환하는 형태로 만들어봐도 좋겠네요 :)

malibinYun commented on app/src/main/java/woowacourse/shopping/domain/repository/HistoryRepository.kt:
여기에만 Result로 반환하지 않는 이유를 다른 개발자들이 의문을 가질거란 생각이 드네요.
DB접근은 Exception이 100% 발생하지 않을까요?

malibinYun commented on app/src/main/java/woowacourse/shopping/view/order/OrderActivity.kt:
```diff
@@ -0,0 +1,126 @@
+import woowacourse.shopping.view.order.adapter.OrderAdapter
+import woowacourse.shopping.view.order.vm.OrderViewModel
+import woowacourse.shopping.view.order.vm.OrderViewModelFactory
+
+class OrderActivity : AppCompatActivity() {
+    private lateinit var networkExceptionDelegator: NetworkExceptionHandler
+
```
여기서 대혼란의 질문을 던지고 도망갈게요.
안드로이드 권장 아키텍처에서 Repository는 반환되는 데이터의 원천을 모르도록 추상회되어있지요.
그렇다면 ViewModel 및 Activity에서는 데이터의 원천에 대한 코드가 없어야한다는 의미로 해석될 수 있는데요,
지금 현재 이 Activity에서 NetworkExceptionHandler라는 객체를 갖고있는 것은 자연스러울까요?

현 코드가 틀렸다는 의미가 아니니 많은 고민을 해보셔도 좋겠어요.

malibinYun commented on app/src/main/java/woowacourse/shopping/view/order/adapter/OrderAdapter.kt:
```diff
@@ -0,0 +1,47 @@
+import woowacourse.shopping.view.core.base.BaseViewHolder
+import woowacourse.shopping.view.order.state.OrderUiState
+
+class OrderAdapter(
+    private val handler: Handler,
+) : ListAdapter<OrderRvItems, BaseViewHolder<ViewBinding>>(OrderDiffUtil()) {
+    override fun getItemViewType(position: Int): Int = getItem(position).viewType.ordinal
```
BaseViewHolder에 다른 구현 타입이 생겨난다면, 다른 개발자들은 이 어뎁터에서 해당 타입도 분기문을 추가해야할지 고민을 가질수도 있겠어요.
모든 타입이 아닌, 이곳에서만 필요한 타입만 사용할 수 있도록 구성해보는 건 어떨까요?

malibinYun reviewed (approved):
제가 최근 일정이 너무 많아 리뷰가 많이 늦었네요. 죄송한 마음이에요.
마지막 미션까지 완주하시느라 정말 고생 많으셨어요.
여러 곳에 코멘트 남겨두었으니 확인해보세요!
지금까지 고민해왔던 것들이 앞으로 개발하면서 큰 도움이 되었으면 해요.
궁금한게 있으면 언제든 연락해도 좋으니 DM 주세요~
